File: app/src/main/java/com/cjstudio/sosestrada/SolicitacaoAdapter.java
package com.cjstudio.sosestrada;

import android.app.AlertDialog;
import android.content.Context;
import android.content.Intent;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.CheckBox;
import android.widget.LinearLayout;
import android.widget.TextView;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import com.google.firebase.firestore.FirebaseFirestore;

import java.util.List;

public class SolicitacaoAdapter extends RecyclerView.Adapter<SolicitacaoAdapter.ViewHolder> {

    private List<Solicitacao> solicitacoes;
    private Context context;
    private FirebaseFirestore db;
    private OnAcaoListener listener;

    public interface OnAcaoListener {
        void onStatusChanged(String id, String novoStatus);
        void onExcluirPermanente(String id);
    }

    public SolicitacaoAdapter(List<Solicitacao> solicitacoes, Context context, OnAcaoListener listener) {
        this.solicitacoes = solicitacoes;
        this.context = context;
        this.db = FirebaseFirestore.getInstance();
        this.listener = listener;
    }

    @NonNull
    @Override
    public ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        View view = LayoutInflater.from(context).inflate(R.layout.item_solicitacao, parent, false);
        return new ViewHolder(view);
    }

    @Override
    public void onBindViewHolder(@NonNull ViewHolder holder, int position) {
        Solicitacao s = solicitacoes.get(position);
        String status = s.getStatus();

        holder.tvMotoristaNome.setText("Motorista: " + s.getMotoristaNome());
        holder.tvMotoristaVeiculo.setText("🚗 Veículo: " + s.getMotoristaVeiculo());
        holder.tvMotoristaPlaca.setText("🔢 Placa: " + s.getMotoristaPlaca());
        holder.tvMotoristaTelefone.setText("📞 Telefone: " + s.getMotoristaTelefone());
        holder.tvMotoristaEndereco.setText("📍 Endereço: " + s.getEnderecoMotorista());

        // Se status for "cancelado", mostra mensagem e esconde os botões
        if ("cancelado".equals(status)) {
            holder.tvCancelamento.setVisibility(View.VISIBLE);
            holder.layoutBotoes.setVisibility(View.GONE);
            holder.btnMensagem.setVisibility(View.GONE);
        } else {
            holder.tvCancelamento.setVisibility(View.GONE);
            holder.layoutBotoes.setVisibility(View.VISIBLE);
            holder.btnMensagem.setVisibility(View.VISIBLE);
        }

        if (s.getNaoLidasPrestador() > 0) {
            holder.btnMensagem.setText("💬 Mensagem (" + s.getNaoLidasPrestador() + " nova"
                    + (s.getNaoLidasPrestador() > 1 ? "s" : "") + ")");
            holder.btnMensagem.setBackgroundTintList(
                    android.content.res.ColorStateList.valueOf(0xFFD32F2F));
        } else {
            holder.btnMensagem.setText("💬 Mensagem");
            holder.btnMensagem.setBackgroundTintList(
                    android.content.res.ColorStateList.valueOf(0xFF9C27B0));
        }

        // Se já foi aceito ou recusado, esconde os botões (mantém comportamento anterior)
        if ("aceito".equals(status) || "recusado".equals(status)) {
            holder.layoutBotoes.setVisibility(View.GONE);
        }

        // Clique em Aceitar → exibe diálogo com checkbox
        holder.btnAceitar.setOnClickListener(v -> {
            if ("pendente".equals(status)) {
                mostrarDialogoConfirmacao(s);
            } else {
                Toast.makeText(context, "Esta solicitação já foi respondida.", Toast.LENGTH_SHORT).show();
            }
        });

        holder.btnRecusar.setOnClickListener(v -> {
            if (listener != null && "pendente".equals(status)) {
                listener.onStatusChanged(s.getId(), "recusado");
            } else {
                Toast.makeText(context, "Esta solicitação já foi respondida.", Toast.LENGTH_SHORT).show();
            }
        });

        holder.btnMensagem.setOnClickListener(v -> {
            Intent intent = new Intent(context, ChatActivity.class);
            intent.putExtra(ChatActivity.EXTRA_SOLICITACAO_ID, s.getId());
            intent.putExtra(ChatActivity.EXTRA_MEU_TIPO, "prestador");
            intent.putExtra(ChatActivity.EXTRA_TITULO, s.getMotoristaNome());
            context.startActivity(intent);
        });

        // Long press para excluir permanentemente (apenas se status for "cancelado")
        holder.itemView.setOnLongClickListener(v -> {
            if ("cancelado".equals(status)) {
                new AlertDialog.Builder(context)
                        .setTitle("Excluir permanentemente")
                        .setMessage("Deseja excluir esta solicitação cancelada?")
                        .setPositiveButton("Sim", (dialog, which) -> {
                            if (listener != null) {
                                listener.onExcluirPermanente(s.getId());
                            }
                        })
                        .setNegativeButton("Não", null)
                        .show();
                return true;
            }
            return false;
        });
    }

    private void mostrarDialogoConfirmacao(Solicitacao s) {
        AlertDialog.Builder builder = new AlertDialog.Builder(context);
        builder.setTitle("Confirmar atendimento");

        // Mensagem + CheckBox
        View view = LayoutInflater.from(context).inflate(R.layout.dialog_aceitar_solicitacao, null);
        CheckBox checkBox = view.findViewById(R.id.checkBoxConfirmacao);
        TextView tvMensagem = view.findViewById(R.id.tvMensagemConfirmacao);

        tvMensagem.setText("Ao aceitar você concorda que estará indo socorrer o motorista, e será enviada uma mensagem ao mesmo, confirmando seu deslocamento.");

        builder.setView(view);

        builder.setPositiveButton("Aceitar", (dialog, which) -> {
            if (checkBox.isChecked()) {
                if (listener != null) {
                    listener.onStatusChanged(s.getId(), "aceito");
                }
            } else {
                Toast.makeText(context, "Marque a checkbox para confirmar.", Toast.LENGTH_SHORT).show();
            }
        });

        builder.setNegativeButton("Cancelar", null);
        builder.show();
    }

    @Override
    public int getItemCount() {
        return solicitacoes.size();
    }

    public void updateList(List<Solicitacao> newList) {
        this.solicitacoes = newList;
        notifyDataSetChanged();
    }

    public static class ViewHolder extends RecyclerView.ViewHolder {
        TextView tvMotoristaNome, tvMotoristaVeiculo, tvMotoristaPlaca, tvMotoristaTelefone, tvMotoristaEndereco, tvCancelamento;
        Button btnAceitar, btnRecusar, btnMensagem;
        LinearLayout layoutBotoes;

        public ViewHolder(@NonNull View itemView) {
            super(itemView);
            tvMotoristaNome = itemView.findViewById(R.id.tvMotoristaNome);
            tvMotoristaVeiculo = itemView.findViewById(R.id.tvMotoristaVeiculo);
            tvMotoristaPlaca = itemView.findViewById(R.id.tvMotoristaPlaca);
            tvMotoristaTelefone = itemView.findViewById(R.id.tvMotoristaTelefone);
            tvMotoristaEndereco = itemView.findViewById(R.id.tvMotoristaEndereco);
            tvCancelamento = itemView.findViewById(R.id.tvCancelamento);
            btnAceitar = itemView.findViewById(R.id.btnAceitar);
            btnRecusar = itemView.findViewById(R.id.btnRecusar);
            layoutBotoes = itemView.findViewById(R.id.layoutBotoes);
            btnMensagem = itemView.findViewById(R.id.btnMensagem);
        }
    }
}

File: app/src/main/java/com/cjstudio/sosestrada/SolicitacaoAdapter.kt
package com.cjstudio.sosestrada

import android.app.AlertDialog
import android.content.Intent
import android.content.res.ColorStateList
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.CheckBox
import android.widget.TextView
import android.widget.Toast
import androidx.recyclerview.widget.RecyclerView
import com.cjstudio.sosestrada.ISolicitacaoRepository.Companion.ACEITO
import com.cjstudio.sosestrada.ISolicitacaoRepository.Companion.CANCELADO
import com.cjstudio.sosestrada.ISolicitacaoRepository.Companion.PENDENTE
import com.cjstudio.sosestrada.ISolicitacaoRepository.Companion.RECUSADO
import com.cjstudio.sosestrada.databinding.ItemSolicitacaoBinding

// Solicitações recebidas pelo prestador (AtendimentoActivity).
class SolicitacaoAdapter(
    private val aoAceitar: (Solicitacao) -> Unit,
    private val aoRecusar: (Solicitacao) -> Unit,
    private val aoExcluir: (Solicitacao) -> Unit
) : RecyclerView.Adapter<SolicitacaoAdapter.ViewHolder>() {

    private var solicitacoes: List<Solicitacao> = emptyList()

    class ViewHolder(val binding: ItemSolicitacaoBinding) : RecyclerView.ViewHolder(binding.root)

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int) =
        ViewHolder(ItemSolicitacaoBinding.inflate(LayoutInflater.from(parent.context), parent, false))

    override fun getItemCount() = solicitacoes.size

    fun atualizarLista(novaLista: List<Solicitacao>) {
        solicitacoes = novaLista
        notifyDataSetChanged()
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        val s = solicitacoes[position]
        val b = holder.binding
        val context = b.root.context

        b.tvMotoristaNome.text = "Motorista: ${s.motoristaNome ?: ""}"
        b.tvMotoristaVeiculo.text = "🚗 Veículo: ${s.motoristaVeiculo ?: ""}"
        b.tvMotoristaPlaca.text = "🔢 Placa: ${s.motoristaPlaca ?: ""}"
        b.tvMotoristaTelefone.text = "📞 Telefone: ${s.motoristaTelefone ?: ""}"
        b.tvMotoristaEndereco.text = "📍 Endereço: ${s.enderecoMotorista ?: "não informado"}"

        val cancelada = s.status == CANCELADO
        b.tvCancelamento.visibility = if (cancelada) View.VISIBLE else View.GONE
        b.btnMensagem.visibility = if (cancelada) View.GONE else View.VISIBLE
        // Aceitar/recusar só enquanto está pendente.
        b.layoutBotoes.visibility = if (s.status == PENDENTE) View.VISIBLE else View.GONE

        if (s.naoLidasPrestador > 0) {
            val plural = if (s.naoLidasPrestador > 1) "s" else ""
            b.btnMensagem.text = "💬 Mensagem (${s.naoLidasPrestador} nova$plural)"
            b.btnMensagem.backgroundTintList = ColorStateList.valueOf(0xFFD32F2F.toInt())
        } else {
            b.btnMensagem.text = "💬 Mensagem"
            b.btnMensagem.backgroundTintList = ColorStateList.valueOf(0xFF9C27B0.toInt())
        }

        b.btnAceitar.setOnClickListener {
            if (s.status == PENDENTE) confirmarAceite(b, s)
            else Toast.makeText(context, "Esta solicitação já foi respondida.", Toast.LENGTH_SHORT).show()
        }
        b.btnRecusar.setOnClickListener {
            if (s.status == PENDENTE) aoRecusar(s)
            else Toast.makeText(context, "Esta solicitação já foi respondida.", Toast.LENGTH_SHORT).show()
        }
        b.btnMensagem.setOnClickListener {
            context.startActivity(
                Intent(context, ChatActivity::class.java)
                    .putExtra(ChatActivity.EXTRA_SOLICITACAO_ID, s.id)
                    .putExtra(ChatActivity.EXTRA_MEU_TIPO, IChatRepository.PRESTADOR)
                    .putExtra(ChatActivity.EXTRA_TITULO, s.motoristaNome)
            )
        }
        // Segurar: exclusão permanente, só de solicitação cancelada.
        b.root.setOnLongClickListener {
            if (!cancelada) return@setOnLongClickListener false
            AlertDialog.Builder(context)
                .setTitle("Excluir permanentemente")
                .setMessage("Deseja excluir esta solicitação cancelada?")
                .setPositiveButton("Sim") { _, _ -> aoExcluir(s) }
                .setNegativeButton("Não", null)
                .show()
            true
        }
    }

    private fun confirmarAceite(b: ItemSolicitacaoBinding, s: Solicitacao) {
        val context = b.root.context
        val view = LayoutInflater.from(context).inflate(R.layout.dialog_aceitar_solicitacao, null)
        val checkBox = view.findViewById<CheckBox>(R.id.checkBoxConfirmacao)
        view.findViewById<TextView>(R.id.tvMensagemConfirmacao).text =
            "Ao aceitar você concorda que estará indo socorrer o motorista, e será enviada uma mensagem ao mesmo, confirmando seu deslocamento."
        AlertDialog.Builder(context)
            .setTitle("Confirmar atendimento")
            .setView(view)
            .setPositiveButton("Aceitar") { _, _ ->
                if (checkBox.isChecked) aoAceitar(s)
                else Toast.makeText(context, "Marque a checkbox para confirmar.", Toast.LENGTH_SHORT).show()
            }
            .setNegativeButton("Cancelar", null)
            .show()
    }

    companion object {
        // Texto mostrado no Toast depois de atualizar o status.
        fun descricaoStatus(status: String) = when (status) {
            ACEITO -> "aceita"
            RECUSADO -> "recusada"
            else -> status
        }
    }
}
